### ISSUE TITLE
Support for JWTs from reusable workflows

### ISSUE THREAD

**Issue by woodruffw** [pypa/warehouse#11096]
Our MVP implementation (#10753) assumes that the workflow is in the same repository, which is [not necessarily true](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/using-openid-connect-with-reusable-workflows).

We should support reusable workflows, specifically via the `job_workflow_ref` claim.

**Comment by woodruffw** [comment]
Just as a note: this will have to interoperate with #11263.

**Comment by GergelyKalmar** [comment]
Any chance to expedite this issue? It seems there's many who would be eager to move over to trusted publishing now that it is the recommended approach, but not being able to do the flow from a reusable workflow is a pretty major blocker. This issue also does not seem to be documented anywhere so we only find out about it once we changed everything to use the new approach.

**Comment by woodruffw** [comment]
> Any chance to expedite this issue? 

I can't personally promise a timeline here, but it is on my backlog of things to do.

> This issue also does not seem to be documented anywhere so we only find out about it once we changed everything to use the new approach.

Thank you for pointing this out; I'll make some changes to the docs to emphasize that reusable workflows are not currently supported.

**Comment by woodruffw** [comment]
> Thank you for pointing this out; I'll make some changes to the docs to emphasize that reusable workflows are not currently supported.

I've opened https://github.com/pypi/warehouse/pull/13592 for this. Thanks again for bringing it to our attention!

**Comment by webknjaz** [comment]
@woodruffw I just stumbled upon an example of a reusable workflow actually _working_ with OIDC here https://github.com/jorisroovers/gitlint/pull/486 — could you confirm this is because of `secrets: inherit` at https://github.com/jorisroovers/gitlint/blob/5247839/.github/workflows/ci.yml#L225 ?

**Comment by woodruffw** [comment]
Yeah, I think that's because it's in the same repo, not because of `secrets: inherit`.

In other words: reusable workflows *do* work with the current implementation, just not reusable workflows that are external to the repository that the trusted publisher was configured with.

**Comment by webknjaz** [comment]
> That workflow is in the same repository though, I think the issue is with reusable workflows that are in a _different_ repository (e.g. in a separate public repo).

Yes, you can't use `inherit` across orgs. The issue was that the OIDC context has a reusable workflow reference and that's currently not checked IIUC. So my concern is that it might be possible to fool the verification logic in PyPI with this...

**Comment by webknjaz** [comment]
I also wonder if it's possible to do something like
```yaml
secrets:
    GITHUB_TOKEN: ${{ github.token }}
```
to make OIDC in the reusable workflow to pretend to be the called workflow...

**Comment by webknjaz** [comment]
> You can get around the `inherit` issue between organizations via forking, which is what we would usually do. So same org, different repo would be fine. The point of reusable workflows in this case would be that you store them in a central repository and you can re-use them in other repositories.

Totally! Though, I've found another use for structuring stuff in-repo: putting different components into separate files while having a centralized workflow that is complex but doesn't have a million lines of YAML in the same file.

**Comment by webknjaz** [comment]
> Just as a note: this will have to interoperate with #11263.

@woodruffw do you have an idea of how this should work in general? Would there be an extra input in the trust form for trusting third-party workflow refs, or how would that need to be implemented?

**Comment by woodruffw** [comment]
Yeah, I've been thinking about this: the "simple" solution would be an extra input like you say, but that has implications for the way we match trusted provider metadata on the backend: every optional configuration point is another query dimension, meaning a linear best, geometric worst increase in queries needed to find the _intended_ publisher configuration.

Rather than doing that, we could relax current "workflow" filename input to accept a reusable workflow reference -- we'd then detect that case, and specialize on it. I need to think more about whether that actually works, but I _think_ it'll be simpler.

**Comment by facutuesca** [comment]
I've been looking into how we can implement this. I've written up the relevant facts about how GitHub and PyPI currently handle OIDC claims, and at the end I specify what adding support for reusable workflows would look like in light of those facts.

### GitHub OIDC claims with (non-)reusable workflows
1. Reusable workflows are workflows that can be called from another workflow
2. There can be [up to 4 levels](https://docs.github.com/en/actions/using-workflows/reusing-workflows#nesting-reusable-workflows) of connected workflows, i.e: `top-level-workflow.yml` -> `reusable-workflow-1.yml` -> `reusable-workflow-2.yml` -> `reusable-workflow-3.yml`.
3. These reusable workflows can be located in external repositories.
4. When GitHub generates an OIDC token, the token will contain claims about the workflow that requested the token.
5. When it's a reusable workflow, the claims will only include information about `top-level-workflow.yml` and `reusable-workflow-3.yml`. That is, only about the top-level and bottom-level workflows.
6. The claims that contain the ref paths for those two workflows are `workflow_ref` and `job_workflow_ref` (top-level and bottom-level respectively). For example:
    ```json
    {
        "workflow_ref": "org/repo/.github/workflows/top-level-workflow.yml@refs/heads/main",
        "job_workflow_ref": "org2/repo2/.github/workflows/reusable-workflow-3.yml@v1",
    }
    ```
7. When the token is requested by a non-reusable workflow (a single workflow that doesn't call other workflows) those two claims will be equal, containing the path ref of that single workflow.
8. There are two more claims for the commit hash of the workflows in the previous claims:
    ```json
    {
        "workflow_sha": "c2698398906ab70572c7878a11debb44ff1d2241",
        "job_workflow_sha": "10040c56a8c0253d69db7c1f26a0d227275512e2",
    }
    ```
9. There is one claim that contains the ref (e.g: `refs/heads/main`) corresponding to the top-level workflow, but not for the bottom-level workflow:
    ```json
    {
        "ref": "refs/heads/main",
    }
    ```

----

### PyPI verification of OIDC claims for GitHub workflows
10. The user creates a Trusted Publisher by specifying a repository owner, a repository name, and a workflow filename (and optionally an environment, which we'll ignore for this discussion since it's not relevant).
11. When PyPI receives an OIDC token, it checks its `job_workflow_ref` claim against the fields mentioned above. For example:

    ```json
    {
        "job_workflow_ref": "org/repo/.github/workflows/release.yml@refs/heads/main",
    }
    ```
    
    will pass validation for a Trusted Publisher configured with:

    ```python
    repository_owner = "org"
    repository_name = "repo"
    workflow_filename = "release.yml"
    ```
12. PyPI will also check that the ref in `jobs_workflow_ref` (in the example above, `refs/heads/main`) matches the `ref` claim, which we saw in (9) corresponds to the top-level workflow.


Note how our checks in (11) and (12) mix claims about the bottom-level workflow (`jobs_workflow_ref`) with claims about the top-level workflow (`ref`). This is fine for non-reusable workflows, since the claims refer all to the same, single workflow. However, it's problematic when starting to consider reusable workflows.

Note also that we accidentally support a subset of reusable workflows: when the top-level and bottom-level workflow are located in the same repo, and the bottom-level workflow is called in the same branch/tag/sha as the top-level one, then the OIDC claims check will pass. Concretely, if the claims look like this:
```json
        "workflow_ref": "org/repo/.github/workflows/top-level-workflow.yml@refs/heads/main",
        "job_workflow_ref": "org/repo/.github/workflows/reusable-workflow-3.yml@refs/heads/main",
        "ref": "refs/heads/main",
```
and the user configured the workflow filename as `reusable-workflow-3.yml`, then verification will pass.


----

### Adding support for reusable workflows

So, what does all of this mean?

First, we want to change the existing OIDC claims verification so that it checks against the top-level workflow instead of the bottom-level one (from `job_workflow_ref` to `workflow_ref`). This should be transparent for most users, but it will break the Trusted Publishers that depend on the (accidental) support for reusable workflows mentioned above. In order to understand how many users this would impact, we will measure how many Trusted Publishers are currently being used with reusable workflows (see PR here: https://github.com/pypi/warehouse/pull/16364).

Second, in order to fully support reusable workflows we need the user to input 3 new pieces of information: the owner and name of the repository where the reusable workflow lives, and the filename of said workflow. Once we have that, we can compare incoming OIDC tokens against the old information (owner, repo and filename of top-level workflow) and against the new information (owner, repo and filename of the bottom-level workflow).

Third, note how existing Trusted Publishers can be trivially migrated by copying the existing 3 fields into the new ones, since for non-reusable workflows they should be equal.

And fourth, from the web UI perspective, the user should not need to fill the 3 new fields when creating a Trusted Publisher for a non-reusable workflow. Rather, those new fields can be hidden behind a "Reusable workflow" checkbox, and when left empty they will default to the same values as the old 3 fields (for the same reason as the point above).

**Comment by mhils** [comment]
Thank you for the detailed write-up, @facutuesca! 😃 🍰 

Is there a reason why we cannot match on `workflow_ref` (the top-level workflow) only? What type of attack are we mitigating by checking `jobs_workflow_ref` as well?

**Comment by facutuesca** [comment]
> Is there a reason why we cannot match on `workflow_ref` (the top-level workflow) only? What type of attack are we mitigating by checking `jobs_workflow_ref` as well?

The scenario we're trying to avoid is having overscoped Trusted Publishers (which is why we suggest having a dedicated, publish-only workflow). In the case of only checking for top-level workflows, imagine you have a project with a top-level `ci.yml` workflow that calls a bunch of other workflows (in the same repo), of which one of them is `publish.yml`:
```
ci.yml -> tests.yml
       -> lint.yml
       -> deploy.yml
       -> build.yml
       -> publish.yml
```

I we only match on `workflow_ref` (that is, `ci.yml`), that means all of the workflows there will be able to publish packages. By adding the extra constraint of `job_workflow_ref`, we can ensure it's only `publish.yml`.

**Comment by webknjaz** [comment]
@facutuesca and the environment name matching remains the same, right?

**Comment by jaraco** [comment]
> Third, note how existing Trusted Publishers can be trivially migrated by copying the existing 3 fields into the new ones, since for non-reusable workflows they should be equal.

For my use-case, I've been holding off using trusted publishing in environments that have reusable workflows until this issue is solved, so I won't have any opinion on the migration path. If you think it would be valuable to employ trusted publishing on such a repo, I'd be open to exploring that and providing feedback on the migration path, but my preference would be to await a durable solution.

**Comment by facutuesca** [comment]
> @facutuesca and the environment name matching remains the same, right?

@webknjaz Yes, the environment check will still be against the `environment` claim sent by GitHub


> > Third, note how existing Trusted Publishers can be trivially migrated by copying the existing 3 fields into the new ones, since for non-reusable workflows they should be equal.
> 
> For my use-case, I've been holding off using trusted publishing in environments that have reusable workflows until this issue is solved, so I won't have any opinion on the migration path. If you think it would be valuable to employ trusted publishing on such a repo, I'd be open to exploring that and providing feedback on the migration path, but my preference would be to await a durable solution.

@jaraco Ah I think I wasn't super clear there. I was referring to the migration from the POV of PyPI: when adding the 3 new columns to the database, we can fill the values with a copy of the existing information (owner, repo, filename). This will be completely transparent to the users, they won't need to migrate or change anything.
I'll reword it to make that point explicit

**Comment by mhils** [comment]
> If we only match on `workflow_ref` (that is, `ci.yml`), that means all of the workflows there will be able to publish packages. By adding the extra constraint of `job_workflow_ref`, we can ensure it's only `publish.yml`.

Thanks, that makes sense! I'm not fully convinced though:

1. The problem you described applies just as much to non-reusable `ci.yml` workflows which may be just as overscoped, no? I don't see an inherent reason why reusable workflows would need additional checks. 
2. Adding an extra checkbox and three new fields adds complexity. Putting the PyPI implementation cost aside for a moment, this makes Trusted Publishing harder to set up for users.
3. If the only goal is to enforce a narrower scope, why not nudge or require the use of environments for new configurations? That helps both non-reusable and reusable workflows. I feel this would be just as effective and would not require users to figure out which workflow path actually goes into which text input.

**Comment by facutuesca** [comment]
> > If we only match on `workflow_ref` (that is, `ci.yml`), that means all of the workflows there will be able to publish packages. By adding the extra constraint of `job_workflow_ref`, we can ensure it's only `publish.yml`.
> 
> Thanks, that makes sense! I'm not fully convinced though:
> 
> 1. The problem you described applies just as much to non-reusable `ci.yml` workflows which may be just as overscoped, no? I don't see an inherent reason why reusable workflows would need additional checks.

True, but I would argue that a parent-level workflow will be overscoped almost by definition, whereas a non-reusable workflow can be made minimally scoped. The argument is: if you only check the top-level `ci.yml` that means that all current _and any future_ workflows called from `ci.yml`, both internal but also from external repos, will be allowed to upload packages.

I think it makes sense there to give the user the ability to narrow it down to a single reusable workflow.

> 2. Adding an extra checkbox and three new fields adds complexity. Putting the PyPI implementation cost aside for a moment, this makes Trusted Publishing harder to set up for users.

True, but only for those who want to set up reusable workflows (the three new fields would be hidden until the user checks the checkbox). So the user experience for the currently supported use case (non-reusable workflows) should be the same, and the added complexity is for the users who want to use the new feature (reusable workflows)

> 3. If the only goal is to enforce a narrower scope, why not nudge or require the use of environments for new configurations? That helps both non-reusable and reusable workflows. I feel this would be just as effective and would not require users to figure out which workflow path actually goes into which text input.

This might be an option. IME though, most projects don't use environments, so making it required would (arguably) be harder for users since they would have to set it up. I'm open to it though, since I might be wrong here.

> Is there anything terribly wrong with changing the check PyPI is currently doing from job_workflow_ref to workflow_ref? My understanding is that reusable workflows would then "just work" while being as secure as non-reusable ones[1](https://github.com/pypi/warehouse/issues/11096#user-content-fn-1-745101dcc45c534c02b0b3604f8992ce).
> The only thing that breaks would be the "accidental support for a subset of reusable workflows" mentioned in

As you mentioned, one issue is breaking currently working Trusted Publishers, which we'll see if it's actually a problem once we have those metrics. The other issue is the discussion we've been having above: if we only check for `workflow_ref`, repos that use reusable workflows will have a very overscoped Trusted Publisher (any workflows called by the top-level workflow, internal or external, current and future, will have permissions to upload).

**Comment by mhils** [comment]
> True, but I would argue that a parent-level workflow will be overscoped almost by definition, whereas a non-reusable workflow can be made minimally scoped. 

Reusable workflows can be made minimally scoped just as well. You may be right that folks are more likely to overscope here, but I don't have any data on this. 🙂 

> This might be an option. IME though, most projects don't use environments, so making it required would (arguably) be harder for users since they would have to set it up. I'm open to it though, since I might be wrong here.

FWIW I think security-wise environments would be the superior solution here. For example, we have lots of folks with write access to the main branch, but only a few selected individuals can approve deployments (and thus push to PyPI).

The one major downside I can think of is that environments are not available for free accounts in private repositories. So that makes it hard to enforce as a strict requirement. If that's a major concern, `job_workflow_ref` may be the better solution.

Thank you for pushing this entire thing forward! 🎉

**Comment by woodruffw** [comment]
> With attestation becoming a more and more important part of the publication process to PyPI I wanted to ask if it would be possible to expedite this work together with #13911.

We're currently working on reusable workflow support. There are no short-term plans (on my part, at least) for #13911, however. That's going to require a larger re-thinking of the data model, since every additional optional field geometrically increases the number of states PyPI needs to check.

**Comment by thepwagner** [comment]
> Is there a reason why we cannot match on workflow_ref (the top-level workflow) only?
> The scenario we're trying to avoid is having overscoped Trusted Publishers 

Could this just be a boolean + warning?

- [ ] Allow reusable workflows.  ⚠ Only enable this if you're using a dedicated release workflow.

Behind the scenes, the checked box would mean only `workflow_ref` is verified, per _mhils suggestion.

---

Or a really out there idea: decode `workflow_ref` and fetch the associated workflow file contents from GitHub.
If you know that contents of `workflow_ref` only specify a single job, then you've mitigated the concern of overscoping without caring about `job_workflow_ref`.
OSSF scorecards use that technique to make assertions about the workflow that is publishing - https://github.com/ossf/scorecard-webapp/blob/main/app/server/verify_workflow.go

**Comment by woodruffw** [comment]
> Could this just be a boolean + warning?

The main reason we can't do this is because, long term, we want to allow both states to enable hermetic CI configurations: a user should be able to configure `pypi.yml` as their "top-level" workflow (i.e. the one in `workflow_ref`), and then a *specific pinned* reusable workflow.

**Comment by jaraco** [comment]
It sounds like the progress here is going in a direction that's not going to help me (and thus trusted publishing will be out of reach indefinitely). I've developed a [system](https://bit.ly/coherent-system) for managing software with the value proposition being that one can maintain packages without having to copy/paste the system-wide configuration (CI config) to each project. So the project uses a minimal project-repo-level config ([example](https://github.com/coherent-oss/coherent.build/blob/main/.github/workflows/main.yml)) that links into a [system-wide config](https://github.com/coherent-oss/system/blob/main/.github/workflows/main.yml) that includes the release hook. So when something needs to change at the system level (such as it [has several times in the past few months](https://github.com/coherent-oss/system/commits/main/.github/workflows)), those changes automatically reflect in the enrolled repos. It wouldn't be suitable for these repos to _pin_ to the shared workflow, because that would defeat the purpose of streamlining the maintenance and would require a commit to each of hundreds of repos each time the system definition needed a tweak.

Do I understand correctly that the planned support for trusted repos for reusable workflows is going to require that the reusable workflow be pinned, and thus won't be suitable for the coherent system?

**Comment by woodruffw** [comment]
> Do I understand correctly that the planned support for trusted repos for reusable workflows is going to require that the reusable workflow be pinned, and thus won't be suitable for the coherent system?

I don't think the plan is to _require_ a pin on the reusable workflow, no -- that's been floated as an idea for _additionally_ constraining the Trusted Publisher (and asserting its hermetic properties), but the basic idea laid out in https://github.com/pypi/warehouse/issues/11096#issuecomment-2260686299 would allow reusable workflows to be configured without a pin, like normal workflow identities currently can be.

**Comment by ekohl** [comment]
For Ruby I ended up implementing a [composite action](https://docs.github.com/en/actions/sharing-automations/creating-actions/creating-a-composite-action): https://github.com/voxpupuli/ruby-release. This works because a composite action doesn't create a new context and you can generate an OIDC token.

**Comment by webknjaz** [comment]
@ekohl it's important to physically separate *jobs* that build things from the job that publishes them. This is because giving elevated privileges to all the transitive deps in the build tool chain is usually a bad idea. Allowing OIDC in other contexts can lead to repo identity impersonation in systems beyond PyPI, which is why I've been actively discouraging that. And additionally, when we were optimizing the action, it ended up being a composite action wrapping a docker one. It turned out that GH doesn't work well with nested composite actions and various unexpected context leak into our env making it defunct. So I've been trying to heavily narrow down the scope of what's considered supported in the action: https://github.com/marketplace/actions/pypi-publish#Non-goals.

That said, I'm planning to make a reusable workflow under PyPA that people would be able to use w/o having to duplicate that within each of their projects. It wouldn't work until PyPI supports this, so I'm not in a hurry to start that just yet. But that workflow will implement what I consider secure practices, so others wouldn't have to think about it and only bring in their jobs that build their dists.

**Comment by ekohl** [comment]
> [@ekohl](https://github.com/ekohl) it's important to physically separate _jobs_ that build things from the job that publishes them.

In my Ruby workflow I use plain `gem build` without installing any dependencies. All I need is `ruby/setup-ruby` which I'd also need for `gem push`. This is different than the Python ecosystem where you have many different implementations.

You can argue that I end up installing `rubygems-await`, which is pulling in some additional dependencies. And I suppose you could also separate the `gem build` into a separate job without credentials since you could run code via the `gemspec`. I'll think about that some more.

**Comment by webknjaz** [comment]
> You can argue that I end up installing `rubygems-await`, which is pulling in some additional dependencies. And I suppose you could also separate the `gem build` into a separate job without credentials since you could run code via the `gemspec`. I'll think about that some more.

I don't know Ruby enough, but it sounds similar to what PEP 517 build backends do. Hence, the same precautions apply. In the reusable workflow I'm thinking of, there'd be exactly 2 tasks — one to download pre-built dists and the second one to upload them. And nothing else, even remotely untrusted.

**Comment by facutuesca** [comment]
Following IRL discussion with @miketheman , @webknjaz and @woodruffw  , here's pseudocode for a potential migration path that adds support for publishing from a (top-level) workflow file that calls another (reusable) workflow file.

```python
claim_top = "workflow_ref"
claim_bottom = "job_workflow_ref"
reusable_workflow_supported_date = "TBD" # date when this change is merged

if claim_top == claim_bottom:  # non-reusable workflow case
    tp_obj = query(TP, TP.filename == claim_top)
    if tp_obj:
        # exchange token
        pass
    else:
        raise TPNotFound()
if claim_top != claim_bottom:  # reusable workflow case
    # if claim_bottom != lastSecurityEvent.bottom: sendEmail()
    # securityEvent.record(TP, claim_bottom)
    tp_obj = query(TP, TP.filename == claim_top)

    if tp_obj:
        # logic for future support of constraining reusable workflow
        # if tp_obj.bottom_filename is None:
        #     # send email (optional)
        #     # exchange token
        #     pass
        # elif tp_obj.bottom_filename == claim_bottom:  # bottom_filename is an optional field
        #     # exchange token
        #     pass
        # elif tp_obj.bottom_filename != claim_bottom:
        #     raise TPNotFound()

        # exchange token
        pass

    else:
        # temporary support for migration purposes
        tp_obj_bottom = query(TP, TP.filename == claim_bottom AND TP.created_date < reusable_workflow_supported_date) 
        if not tp_obj_bottom:
            raise TPNotFound()
        else:
            # exchange token
            # send email about TP referring to reusable instead of top-level workflow
            pass
```

**Comment by webknjaz** [comment]
@facutuesca do you remember, when talking about the UI changes needed in May, did we discuss the GitHub Environment field? I recall we thought that a checkbox might show a few extra fields for pointing to the reusable workfow but I think the calling side might not allow setting an env in GHA (needs veification) and then, only the called side would be the one with the environment. We might need to check this bit.

**Comment by carlcsaposs-canonical** [comment]
> [@facutuesca](https://github.com/facutuesca) do you remember, when talking about the UI changes needed in May, did we discuss the GitHub Environment field? I recall we thought that a checkbox might show a few extra fields for pointing to the reusable workfow but I think the calling side might not allow setting an env in GHA (needs veification) and then, only the called side would be the one with the environment. We might need to check this bit.

Not sure if this is helpful, but in case this is what you're referring to:

I believe the reusable workflow can set the environment (e.g. https://github.com/canonical/data-platform-workflows/blob/af081177f7b6bf38f751f80ff5dc464950c047d0/.github/workflows/release_charm_edge.yaml#L47) and it's used in the context of the caller repository

I don't think the calling workflow can set the environment for a reusable workflow, but I think it could pass an input to the reusable workflow that the reusable workflow could use to set the environment

**Comment by artpelling** [comment]
Could someone give an update on this issue? Is support for reusable workflows planned? If yes, roughly until when?

**Comment by webknjaz** [comment]
> Could someone give an update on this issue?

Unlikely.

> Is support for reusable workflows planned?

It is known that people want this and nobody's against accepting a correctly implemented contribution.

> If yes, roughly until when?

No timeline, this fully depends on the volunteer effort at the moment. You can help this to happen by convincing your employer to fund the work. Maybe they could help the @psf hire @facutuesca through the @trailofbits, for example.

**Comment by lmmx** [comment]
In case it might unblock anyone reading this thread, I just wanted to point out a recent approach to workaround this issue: disable PyPI upload in the reusable workflow, instead upload package artifacts then pass to `pypa/gh-action-pypi-publish`

- See docs at https://github-actions-workflows.openastronomy.org/en/latest/trusted_publishing.html 
  - via: https://github.com/OpenAstronomy/github-actions-workflows/issues/136#issuecomment-3800209172

**Comment by webknjaz** [comment]
@lmmx this is kinda the only recommended way right now..
